keps/sig-apps/20181220-statefulset-volume-expansion.md

```diff
@@ -0,0 +1,118 @@
+replaces:
+  - n/a
+superseded-by:
+  - n/a
+---
+
+# StatefulSet Volume Expansion
+
+## Table of Contents
+
+* [Table of Contents](#table-of-contents)
+* [Summary](#summary)
+* [Motivation](#motivation)
+    * [Goals](#goals)
+* [Proposal](#proposal)
+    * [Implementation Details/Notes/Constraints](#implementation-detailsnotesconstraints)
+    * [Risks and Mitigations](#risks-and-mitigations)
+* [Graduation Criteria](#graduation-criteria)
+* [Implementation History](#implementation-history)
+
+[Tools for generating]: https://github.com/ekalinin/github-markdown-toc
+
+## Summary
+
+The purpose of this enhancement is to allow for the expansion of persistent volume claims created by StatefulSets. This entails propagating increases to storage requests in `StatefulSets.volumeClaimTemplates` to associated persistent volume claims.
+
+## Motivation
+In Kubernetes v1.11 the persistent volume expansion feature was promoted to beta. This allowed users to expand volumes by editing storage requests in persistent volume claim objects.
+
+Kubernetes creates a persistent volume for each `volumeClaimTemplate` in the `volumeClaimTemplates` component of a StatefulSet. However, it is not possible to expand persistent volumes, created by StatefulSets, by editing the source `volumeClaimTemplate` in the StatefulSet object. Therefore, it is necessary for the user to individually modify all pods' persistent volume claims, by increasing their storage requests, to expand the underlying persistent volumes. This would have to be repeated each time the number of replicas in the StatefulSet object is increased, since new persistent volumes would be created with the original storage request specified in `volumeClaimTemplates` component.
+
+It would be easier and expected to allow for changes to storage requests in the `volumeClaimTemplates` component of a StatefulSet to propagate to all associated persistent volume claims.
+
+Relevant Issues:
+
+* https://github.com/kubernetes/kubernetes/issues/71477
+* https://github.com/kubernetes/kubernetes/issues/72198
+
+### Goals
+
+Allow for increases to storage requests in the `volumeClaimTemplates` component of a StatefulSet to propagate to all associated persistent volume claims.
+
+## Proposal
+
+### Implementation Details/Notes/Constraints
+
+The apiserver will allow for increases to storage requests in the `volumeClaimTemplates` component of a StatefulSet. Additionally, for each volumeClaimTemplate being expanded, it will be necessary to check if its associated StorageClass has volume expansion enabled. This will be achieved by updating the `PersistentVolumeClaimResize` admission controller to incorporate validating updates to `volumeClaimTemplates` within a StatefulSet. Specifically the admission controller will now also check if every `volumeClaimTemplate` in a StatefulSet, which is being resized, is associated with a StorageClass that has `allowVolumeExpansion = true`.
```

Comment:
> This will be achieved by updating the `PersistentVolumeClaimResize` admission controller to incorporate validating updates to `volumeClaimTemplates` within a StatefulSet.

This is somewhat unusual... we don't typically look at all template objects in admission (admission plugins that look at pods don't typically look at replicaset/statefulset/deployment/job/cronjob templates, pvc admission plugins don't typically look at statefulset templates).

Do we currently gate statefulset creation on existence of the referenced storageclass?

Reply:
This also means that the operations of updating a storageclass to be resizeable and requesting a size increase in a statefulset are order-dependent. Is that what we want?

What is the behavior if we allowed requesting a size increase in a statefulset that could not be fulfilled because the storage class was not resizeable? (even with the admission check, it's possible to modify a resizable storageclass and mark it not resizeable, so the statefulset controller still needs to handle the case where resizing the PVC is not allowed)

Reply:
> Do we currently gate statefulset creation on existence of the referenced storageclass?

No the creation of a statefulset is not gated on the existence of the referenced storageclass.

> This also means that the operations of updating a storageclass to be resizeable and requesting a size increase in a statefulset are order-dependent. Is that what we want?

I think what we want (similar to what you mentioned above) is to try our best to ensure that the statefuset controller does not allow a volumeClaimTemplate storage request to increase if we are certain it will fail. So we want to atleast make sure that the storageClass is marked as resizeable before we accept the storage request increase. 

> What is the behavior if we allowed requesting a size increase in a statefulset that could not be fulfilled because the storage class was not resizeable? 

Now if the statefulset controller encounters the case where the PVC resize request could not be fulfilled either at the admission layer(storageclass marked not resizeable in between statfulset spec update and the resize call to pvc) or due to the resize controller failing, the documented workaround in the [kep](https://github.com/kubernetes/enhancements/pull/660/files#diff-60237b8b1acaebbcb71b1ea3526a42c0R86)is the suggested option. (Note this is based off of the documented workaround to recover from a failing volume expansion in general). So in this case deleting and recreating the statefulset set would suffice (pvcs remain so no data loss). A rollback would be possible if the volumeClaimTemplate was not in the controller revision as you mentioned below.

Reply:
There will be always be a chance that storageclass field does not reflect true capability of the volume plugin. for in-tree drivers this is easy and each plugin can be queried. For CSI plugins, we will have to rely on `CSIDriver` object but we weren't considering making `CSIDriver` a necessity for CSI volume expansion of stateful sets. 

So the worst this could do is - it will cause all PVCs managed by stateful set to be stuck in "Resizing" condition. The resize will not actually be retried because resize controller knows that plugin itself does not support resize. The PVCs themselves will be usable and nothing stops them from being used in pod/statefulsets if they have "Resizing" condition.

Reply:
But I see how this interacts with @liggitt's comment below and could prevent an user from rolling back to revision of StatefulSet.

Reply:
As mentioned above and in the comment [below](https://github.com/kubernetes/enhancements/pull/660#discussion_r325235412),  including the requested volume size in the revision prevents the statefulset from being rolled back (decreases to volume requests are rejected at the api). However, the controller logic for updating pods is centred around the revision name. Thus if the storage request is not included in the revision the update is not triggered per the usual control flow as the revision name does not change. 

Off the top of my head we could probably change another field in the spec which would also be included in the revision and could be rolled back. This would be changed along with any increase to a storage request. This way the controller update logic would remain unchanged and rollbacks would ignore storageRequests. On the other hand including a new field just for this purpose doesn't seem to be a good idea...

We could also mark the pod for termination if the requested size > actual size (for the pvc). This would be another condition (along with change is revision) that would trigger a pod to be deleted for update. Not sure if this would break any assumptions since updates seem to be based around changes to revision names. 

@janetkuo @gnufied thoughts?